File: JAVA/level2/압축.java
import java.util.*;

class Solution {
    public int[] solution(String msg) {
        int[] answer = {};
        HashMap<String, Integer> map = new HashMap<String, Integer>();
        ArrayList<Integer> temp_answer = new ArrayList<Integer>();
        String[] characters = {"A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"};
        int index = 27;
        String current = Character.toString(msg.charAt(0));
        String next = "";

        for(int i=0; i<26; i++)
            map.put(characters[i], i+1);

        for(int i=1; i<msg.length(); i++){
            next = Character.toString(msg.charAt(i));

            if(map.containsKey(current+next))
                current += next;
            else{
                temp_answer.add(map.get(current));
                map.put(current+next, index++);
                current = Character.toString(msg.charAt(i));
            }
        }

        temp_answer.add(map.get(current));

        answer = new int[temp_answer.size()];

        for(int i=0; i<temp_answer.size(); i++)
            answer[i] = temp_answer.get(i);

        return answer;
    }
}

File: JavaScript/level2/압축.js
function solution(msg) {
    var answer = [];
    let indexMap = {A:1, B:2, C:3, D:4, E:5, F:6, G:7, H:8, I:9, J:10, K:11, L:12, M:13, N:14, O:15, P:16, Q:17, R:18, S:19, T:20, U:21, V:22, W:23, X:24, Y:25, Z:26}
    let current = msg[0];
    let next = "";
    let index = 27;
    for(let i=1; i<msg.length; i++){
        next = msg[i];
        
        if(current+next in indexMap)
            current = current+next;
        else{
            answer.push(indexMap[current]);
            indexMap[current + next] = index++;
            current = next;
        }
    }
    answer.push(indexMap[current]);

    return answer;
}